Source/DafnyCore/AST/DafnyAst.cs

```diff
@@ -31,6 +33,15 @@ public interface INode {
     /// losing source location information, so those transformed nodes should not be returned by this property.
     /// </summary>
     IEnumerable<INode> Children { get; }
+
+    /// <summary>
+    /// These children should match what was parsed before the resolution phase.
+    /// That way, gathering all OwnedTokens of all recursive ConcreteChildren should result in a comprehensive
+    /// coverage of the original program
+    /// </summary>
+    IEnumerable<INode> ConcreteChildren { get; }
+
+    List<IToken> OwnedTokens { get; }
```

Comment:
> Why is that?
> What does TopLevel mean in the name GetFirstTopLevelToken ? Have you considered emphasising that it's getting the first token that's actually in the file, like with GetFirstTokenInsideFile ?

This method was originally implemented on modules to report errors on the module itself when tokens were not available (i.e. a crash). Then, this method was implemented at the program level because there could be crashes without modules.
I added a comment to `GetFirstTopLevelToken` to explain what it extracts in the case of a file.